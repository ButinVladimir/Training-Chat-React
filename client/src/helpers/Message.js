import PropTypes from 'prop-types';

class Message {
  constructor(id, date, from, to, message) {
    this.id = id;
    this.date = date;
    this.from = from;
    this.to = to;
    this.message = message;
  }
}

const messageShape = {
  id: PropTypes.string.isRequired,
  date: PropTypes.string.isRequired,
  from: PropTypes.string.isRequired,
  to: PropTypes.arrayOf(PropTypes.string),
  message: PropTypes.string.isRequired,
};

export default Message;
export { messageShape };
