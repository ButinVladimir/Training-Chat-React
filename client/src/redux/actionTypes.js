export const CHANGE_LOGIN = 'CHANGE_LOGIN';
export const LOGIN = 'LOGIN';
export const LOGOUT = 'LOGOUT';
export const CHANGE_SPEECH = 'CHANGE_SPEECH';
export const SAY = 'SAY';
export const ADD_TO = 'ADD_TO';
export const REMOVE_TO = 'REMOVE_TO';
export const CONNECT = 'CONNECT';
export const DISCONNECT = 'DISCONNECT';
export const SHOW_ERROR = 'SHOW_ERROR';
export const HIDE_ERROR = 'HIDE_ERROR';
export const USER_APPEARS = 'USER_APPEARS';
export const USER_DISAPPEARS = 'USER_DISAPPEARS';
export const GET_MESSAGE = 'GET_MESSAGE';
export const UPDATE_USER_LIST = 'UPDATE_USER_LIST';
