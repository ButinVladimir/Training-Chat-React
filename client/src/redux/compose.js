export default function compose(...args) {
  return (state, action) => {
    let resultState = state;
    args.forEach((reducer) => { resultState = reducer(resultState, action); });

    return resultState;
  };
}
