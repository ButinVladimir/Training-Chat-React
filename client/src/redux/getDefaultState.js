export default () => ({
  loginPanel: {
    loginned: false,
    login: '',
  },
  usersPanel: {
    usersList: [],
  },
  speechPanel: {
    speech: '',
    to: [],
  },
  chatPanel: {
    messages: [],
  },
  connected: false,
  errorPanel: {
    showError: false,
    error: '',
  },
});
