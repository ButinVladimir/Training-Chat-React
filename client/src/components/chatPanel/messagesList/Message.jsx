import React from 'react';
import PropTypes from 'prop-types';
import { messageShape } from '../../../helpers/Message';

function Message({
  id,
  date,
  from,
  to,
  message,
  onClickUser,
}) {
  const convertedTo = to
    ? to.map(t => (
      <button
        type="button"
        key={t}
        className="user-select to"
        onClick={() => onClickUser(t)}
      >
        {`@${t}`}
      </button>))
    : [];

  return (
    <li className="chat-message" key={id}>
      <span className="date">{`[${date}]`}</span>
      <button
        type="button"
        className="user-select from"
        onClick={() => onClickUser(from)}
      >
        {`${from}: `}
      </button>
      {convertedTo}
      <span className="message">{message}</span>
    </li>
  );
}

Message.propTypes = {
  ...messageShape,
  onClickUser: PropTypes.func.isRequired,
};

export default Message;
