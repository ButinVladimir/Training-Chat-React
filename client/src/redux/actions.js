import * as actionTypes from './actionTypes';

export const changeLogin = loginValue => ({
  type: actionTypes.CHANGE_LOGIN,
  loginValue,
});

export const login = () => ({
  type: actionTypes.LOGIN,
});

export const logout = () => ({
  type: actionTypes.LOGOUT,
});

export const changeSpeech = speechValue => ({
  type: actionTypes.CHANGE_SPEECH,
  speechValue,
});

export const say = (id, date) => ({
  type: actionTypes.SAY,
  id,
  date,
});

export const addTo = recipient => ({
  type: actionTypes.ADD_TO,
  recipient,
});

export const removeTo = recipient => ({
  type: actionTypes.REMOVE_TO,
  recipient,
});
