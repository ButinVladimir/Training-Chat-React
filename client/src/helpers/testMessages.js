import Message from './Message';

const testMessages = [
  new Message('1', '8/28/2018', 'User 1', [], 'Message 1'),
  new Message('2', '8/28/2018', 'User 1', [], 'Message 2'),
  new Message('3', '8/28/2018', 'User 1', ['User 2'], 'Message 3 to one user'),
  new Message('4', '8/28/2018', 'User 2', ['User 1', 'User 2'], 'User 2 answer'),
  new Message('5', '8/28/2018', 'User 3', ['User 1', 'User 2'], 'User 3 question?'),
  new Message('6', '8/28/2018', 'User 1', ['User 1', 'User 2', 'User 3'], 'User 1 answer'),
  new Message('7', '8/28/2018', 'User 2', [], 'Message 7'),
  new Message('8', '8/28/2018', 'User 2', [], 'Message 8'),
  new Message('9', '8/28/2018', 'User 1', [], 'Message 9'),
  new Message('10', '8/28/2018', 'User 3', [], 'Message 10'),
  new Message('11', '8/28/2018', 'User 1', [], 'Message 1'),
  new Message('12', '8/28/2018', 'User 1', [], 'Message 2'),
  new Message('13', '8/28/2018', 'User 1', ['User 2'], 'Message 3 to one user'),
  new Message('14', '8/28/2018', 'User 2', ['User 1', 'User 2'], 'User 2 answer'),
  new Message('15', '8/28/2018', 'User 3', ['User 1', 'User 2'], 'User 3 question?'),
  new Message('16', '8/28/2018', 'User 1', ['User 1', 'User 2', 'User 3'], 'User 1 answer'),
  new Message('17', '8/28/2018', 'User 2', [], 'Message 7'),
  new Message('18', '8/28/2018', 'User 2', [], 'Message 8'),
  new Message('19', '8/28/2018', 'User 1', [], 'Message 9'),
  new Message('20', '8/28/2018', 'User 3', [], 'Message 10'),
  new Message('21', '8/28/2018', 'User 1', [], 'Message 1'),
  new Message('22', '8/28/2018', 'User 1', [], 'Message 2'),
  new Message('23', '8/28/2018', 'User 1', ['User 2'], 'Message 3 to one user'),
  new Message('24', '8/28/2018', 'User 2', ['User 1', 'User 2'], 'User 2 answer'),
  new Message('25', '8/28/2018', 'User 3', ['User 1', 'User 2'], 'User 3 question?'),
  new Message('26', '8/28/2018', 'User 1', ['User 1', 'User 2', 'User 3'], 'User 1 answer'),
  new Message('27', '8/28/2018', 'User 2', [], 'Message 7'),
  new Message('28', '8/28/2018', 'User 2', [], 'Message 8'),
  new Message('29', '8/28/2018', 'User 1', [], 'Message 9'),
  new Message('30', '8/28/2018', 'User 3', [], 'Message 10'),
];

export default testMessages;
