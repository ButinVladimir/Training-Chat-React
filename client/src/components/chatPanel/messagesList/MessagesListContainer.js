import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Socket } from 'socket.io-client';
import MessagesList from './MessagesList';
import { getMessage } from '../../../redux/actions';
import { GET_MESSAGE } from '../../../helpers/socketEvents';
import getId from '../../../helpers/getId';
import getDate from '../../../helpers/getDate';

const mapStateToProps = state => ({
  messages: state.chatPanel.messages,
});

const mapDispatchToProps = (dispatch, ownProps) => {
  const handler = (message, from, to) => {
    dispatch(getMessage(message, from, to, getId(), getDate()));
  };

  ownProps.socket.on(GET_MESSAGE, handler);

  return {
    handlers: new Map([[GET_MESSAGE, handler]]),
  };
};

const MessagesListContainer = connect(mapStateToProps, mapDispatchToProps)(MessagesList);

MessagesListContainer.propTypes = {
  socket: PropTypes.instanceOf(Socket).isRequired,
};

export default MessagesListContainer;
