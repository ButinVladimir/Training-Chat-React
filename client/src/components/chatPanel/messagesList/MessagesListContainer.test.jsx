import React from 'react';
import { Socket } from 'socket.io-client';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import MessagesListContainer from './MessagesListContainer';
import Message from '../../../helpers/Message';
import { GET_MESSAGE as SOCKET_GET_MESSAGE } from '../../../helpers/socketEvents';
import { GET_MESSAGE as REDUX_GET_MESSAGE} from '../../../redux/actionTypes';

describe('MessageContainer', () => {
  let mockStore;
  let mockSocket;

  beforeEach(() => {
    mockStore = configureStore([]);
    mockSocket = new Socket();
  });

  it('renders list of messages', () => {
    const store = mockStore({
      chatPanel: {
        messages: [
          new Message('id', 'date', 'from', ['to1', 'to2'], 'message'),
        ],
      },
    });
    const component = mount(
      <Provider store={store}>
        <MessagesListContainer
          socket={mockSocket}
        />
      </Provider>,
    );

    const messages = component.find('MessagesList').find('Message');
    expect(component).toMatchSnapshot();
    expect(messages).toHaveLength(1);
    expect(messages.at(0).props().id).toEqual('id');
  });

  it('handles getting message', () => {
    const store = mockStore({
      chatPanel: {
        messages: [],
      },
    });
    const component = mount(
      <Provider store={store}>
        <MessagesListContainer
          socket={mockSocket}
        />
      </Provider>,
    );

    let actions = store.getActions();
    let expectedActions = [];

    expect(actions).toEqual(expectedActions);
    expect(mockSocket.__mockHandlers.has(SOCKET_GET_MESSAGE)).toEqual(true);
    mockSocket.__mockHandlers.get(SOCKET_GET_MESSAGE)('message', 'from', [], '', '');

    actions = store.getActions();
    expectedActions = [{ type: REDUX_GET_MESSAGE, message: 'message', from: 'from', to: []}];
    expect(actions).toMatchObject(expectedActions);
  });
});
