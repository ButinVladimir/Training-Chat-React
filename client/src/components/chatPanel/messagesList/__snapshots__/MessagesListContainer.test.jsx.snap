// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageContainer renders list of messages 1`] = `
<div
  className="chat-messages-list-container"
>
  <ul
    className="chat-messages-list"
  >
    <li
      className="chat-message"
      key="id"
    >
      <span
        className="date"
      >
        [date]
      </span>
      <button
        className="user-select from"
        onClick={[Function]}
        type="button"
      >
        from: 
      </button>
      <button
        className="user-select to"
        key="to1"
        onClick={[Function]}
        type="button"
      >
        @to1
      </button>
      <button
        className="user-select to"
        key="to2"
        onClick={[Function]}
        type="button"
      >
        @to2
      </button>
      <span
        className="message"
      >
        message
      </span>
    </li>
  </ul>
</div>
`;
